Automatically handle IBus configuration to avoid conflicts with KDE

ArrayBolt3 opened the pull request:
Resolves the issue in internal ticket 4027. No user intervention should be needed beyond following a pop-up prompt that may appear on login for people with a misconfigured IBus installation.

ArrayBolt3 committed:
4027 automatically handle IBus configuration to avoid conflicts with KDE

mmikowski commented on package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:
```diff
@@ -0,0 +1,104 @@
+
+_fi2StderrFn () { 1>&2 echo -e  "$*"; }
+
+## BEGIN _fi2RepairIbusConfigFn {
```
Fantastic documentation. If we create a rtest, we should move all this there. Add a comment to move to rtest when the time is right.

writing a rtest is actually quite simple for this; can mock qdbus...

mmikowski commented on package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:
```diff
@@ -0,0 +1,104 @@
+
+  _mode_str="${1:-}";
+
+  if [ "${_mode_str}" = 'check' ]; then
```
1. Provide a mode that checks and then applies if needed. 
2. For multi-branch checks, use case statements

```bash
case "${_mode_str}" in
  check) ;;
  repair) ;;
  *) ;; 
esac

mmikowski commented on package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:
```diff
@@ -0,0 +1,104 @@
+## BEGIN _fi2MainFn {
+_fi2MainFn () {
+  declare _mode_str;
+
```
```bash
if [ -z "${_qdbusExe}" ]; then
  # Avoid pipelines whenever you can. This is almost certainly better like this:
  if grep -qa 'org.freedesktop.ibus' < <(qdbus); then
    _fi2StderrFn 'qdbus not found';
    return 1;
  fi
```
...

mmikowski commented on package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:
```diff
@@ -0,0 +1,104 @@
+  _binDir="$(   dirname  "${_binName}" )" || exit 101;
+  _baseDir="$(  dirname  "${_binDir}"  )" || exit 101;
+  _baseName="$( basename "${_binName}" )" || exit 101;
+
```
```
_qdbusExe="$(command -v qdbus)";
```

mmikowski commented on package-main/usr/lib/kfocus/bin/kfocus-fix-ibus:
```diff
@@ -0,0 +1,104 @@
+  fi
+}
+## . END _fi2MainFn }
+
```
Add declaration of package variables here
```bash
declare _binName  ...

mmikowski reviewed (changes requested):
Looks excellent, but could you please make the few tweaks and then just let me know.  I'd appreciate it!

mmikowski committed:
4027 Add regression test, Please move Docs here!

ArrayBolt3 committed:
4027 improve regression tests

ArrayBolt3 committed:
4027 regression test improvements, generate expect data

ArrayBolt3 commented:
I believe I have implemented the requested changes. I also adjusted the main worker script to make it easier to test, and got the regression tests in much better shape.

ArrayBolt3 commented:
I can't figure out how to get this marked as ready for review :-/ :P

ArrayBolt3 committed:
4027 final cleanups, adjust expect data format
